test/kubemark/cloudprovider.go

```diff
@@ -0,0 +1,342 @@
+	node.Name = fmt.Sprintf("%s-%d", nodeGroup, rand.Int63())
+	node.Labels = map[string]string{nodeGroupLabel: nodeGroup, "name": node.Name}
+	node.Spec.Template.Labels = node.Labels
+	_, err = kubemarkProvider.externalCluster.client.CoreV1().ReplicationControllers(node.Namespace).Create(node)
+	return err
+}
+
+// getNodeTemplate returns the template for hollow node replication controllers
+// by looking for an existing replication controller with a specific name. This
+// requires at least one kubemark node to be present on startup.
+func (kubemarkProvider *Provider) getNodeTemplate() (*apiv1.ReplicationController, error) {
+	if hollowNode := kubemarkProvider.getReplicationControllerByName(hollowNodeName); hollowNode != nil {
+		nodeTemplate := &apiv1.ReplicationController{
+			Spec: apiv1.ReplicationControllerSpec{
+				Template: hollowNode.Spec.Template,
+			},
+		}
+
+		nodeTemplate.Spec.Selector = nil
+		nodeTemplate.Namespace = namespaceKubemark
+		one := int32(1)
+		nodeTemplate.Spec.Replicas = &one
+
+		return nodeTemplate, nil
+	}
+	return nil, fmt.Errorf("Can't get hollow node template.")
+}
+
+func (kubemarkCluster *KubemarkCluster) removeUnneededNodes(oldObj interface{}, newObj interface{}) {
+	node, ok := newObj.(*apiv1.Node)
+	if !ok {
+		return
+	}
+	for _, condition := range node.Status.Conditions {
+		// Delete node if it is in unready state, and it has been
+		// explicitly marked for deletion.
+		if condition.Type == apiv1.NodeReady && condition.Status != apiv1.ConditionTrue {
+			kubemarkCluster.nodesToDeleteLock.Lock()
+			defer kubemarkCluster.nodesToDeleteLock.Unlock()
+			if kubemarkCluster.nodesToDelete[node.Name] {
+				kubemarkCluster.nodesToDelete[node.Name] = false
+				if err := kubemarkCluster.client.CoreV1().Nodes().Delete(node.Name, &metav1.DeleteOptions{}); err != nil {
+					glog.Errorf("Failed to delete node %s from kubemark cluster", node.Name)
+				}
+			}
+			return
+		}
+	}
+}
+
+func (kubemarkCluster *KubemarkCluster) markNodeForDeletion(name string) {
+	kubemarkCluster.nodesToDeleteLock.Lock()
+	defer kubemarkCluster.nodesToDeleteLock.Unlock()
+	kubemarkCluster.nodesToDelete[name] = true
+}
+
+func newReplicationControllerInformer(kubeClient kube_client.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
+	rcListWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "replicationcontrollers", namespaceKubemark, fields.Everything())
+	return cache.NewSharedIndexInformer(rcListWatch, &apiv1.ReplicationController{}, resyncPeriod, nil)
+}
+
+func newPodInformer(kubeClient kube_client.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
+	podListWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", namespaceKubemark, fields.Everything())
+	return cache.NewSharedIndexInformer(podListWatch, &apiv1.Pod{}, resyncPeriod, nil)
+}
+
+func newNodeInformer(kubeClient kube_client.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
```

Comment:
We don't need this logic and can get rid of the function. The informer can directly be obtained from the factory as mentioned in https://github.com/kubernetes/kubernetes/pull/49382/files#r129030754.